src/main/io/gimbal_control.c

```diff
@@ -0,0 +1,224 @@
+        return false;
+    }
+
+    // Scale the incoming ±500 range to the max values accepted by the gimbal
+    roll = ((GIMBAL_ROLL_MAX + GIMBAL_ROLL_MIN) / 2) + ((GIMBAL_ROLL_MAX - GIMBAL_ROLL_MIN) * roll / 1000);
+    pitch = ((GIMBAL_PITCH_MAX + GIMBAL_PITCH_MIN) / 2) - ((GIMBAL_PITCH_MAX - GIMBAL_PITCH_MIN) * pitch / 1000);
+    yaw = ((GIMBAL_YAW_MAX + GIMBAL_YAW_MIN) / 2) + ((GIMBAL_YAW_MAX - GIMBAL_YAW_MIN) * yaw / 1000);
+
+    // Constrain to catch any incoming out of range values
+    gimbalCmd.roll = constrain(roll, GIMBAL_ROLL_MIN, GIMBAL_ROLL_MAX);
+    gimbalCmd.pitch = constrain(pitch, GIMBAL_PITCH_MIN, GIMBAL_PITCH_MAX);
+    gimbalCmd.yaw = constrain(yaw, GIMBAL_YAW_MIN, GIMBAL_YAW_MAX);
+
+    DEBUG_SET(DEBUG_GIMBAL, 3, gimbalCmd.roll);
+    DEBUG_SET(DEBUG_GIMBAL, 4, gimbalCmd.pitch);
+    DEBUG_SET(DEBUG_GIMBAL, 5, gimbalCmd.yaw);
+
+    gimbalCmd.mode = gimbalTrackConfig()->gimbal_stabilisation;
+    gimbalCmd.sens = gimbalTrackConfig()->gimbal_sensitivity;
+
+    uint16_t crc = gimbalCrc((uint8_t *)&gimbalCmd, sizeof(gimbalCmd) - 2);
+
+    gimbalCmd.crch = (crc >> 8) & 0xff;
+    gimbalCmd.crcl = crc & 0xff;
+
+    return true;
+}
+
+// Set the gimbal position on each axis in a ±500 range with scale and offset applied
+bool gimbalTrack(int16_t throttle, int16_t roll, int16_t pitch, int16_t yaw)
+{
+    int16_t gimbal_roll;
+    int16_t gimbal_pitch;
+    int16_t gimbal_yaw;
+
+    // Apply scale an offset, each of which are expressed as a percentage
+    gimbal_roll = (roll * gimbalTrackConfig()->gimbal_roll_gain / 100) + 5 * gimbalTrackConfig()->gimbal_roll_offset;
+    gimbal_yaw = (yaw * gimbalTrackConfig()->gimbal_yaw_gain / 100) + 5 * gimbalTrackConfig()->gimbal_yaw_offset;
+
+    if (pitch < 0) {
+        gimbal_pitch = pitch * gimbalTrackConfig()->gimbal_pitch_low_gain / 100;
+    } else {
+        gimbal_pitch = pitch * gimbalTrackConfig()->gimbal_pitch_high_gain / 100;
+    }
+
+    gimbal_pitch += 5 * gimbalTrackConfig()->gimbal_pitch_offset - (throttle * gimbalTrackConfig()->gimbal_pitch_thr_gain / 100);
+
+    return gimbalSet(gimbal_roll, gimbal_pitch, gimbal_yaw);
+}
+
+// Gimbal updates should be sent at 100Hz or the gimbal will self center after approx. 2 seconds
+void gimbalUpdate(timeUs_t currentTimeUs)
+{
+    UNUSED(currentTimeUs);
+
+    if (!gimbalSerialPort) {
+        setTaskEnabled(TASK_GIMBAL, false);
+        return;
+    }
+    serialWriteBuf(gimbalSerialPort, (uint8_t *)&gimbalCmd, sizeof(gimbalCmd));
+}
+
+bool gimbalInit(void)
+{
+    const serialPortConfig_t *portConfig = findSerialPortConfig(FUNCTION_GIMBAL);
+    if (!portConfig) {
+        return false;
+    }
+
+    // Serial communications is 115200 8N1
+    gimbalSerialPort = openSerialPort(portConfig->identifier, FUNCTION_GIMBAL,
+                                      NULL, NULL,
+                                      115200, MODE_RXTX, SERIAL_STOPBITS_1 | SERIAL_PARITY_NO);
+
+    gimbalCmd.opcode[0] = GIMBAL_CMD_S0;
```

Comment:
This may be moved into gimbalSet

Reply:
It's where it is as it only needs to be done once.